src/NSubstitute/Core/Arguments/ArrayArgumentSpecificationsFactory.cs

```diff
@@ -21,7 +22,15 @@ public IEnumerable<IArgumentSpecification> Create(object arrayArgument, IEnumera
             var result = new List<IArgumentSpecification>();
             foreach (var member in arrayMembers)
             {
-                result.Add(_nonParamsArgumentSpecificationFactory.Create(member, parameterInfos.ElementAt(index), suppliedArgumentSpecifications)); 
+                try
+                {
+                    result.Add(_nonParamsArgumentSpecificationFactory.Create(member, parameterInfos.ElementAt(index), suppliedArgumentSpecifications));
+                }
+                catch (AmbiguousArgumentsException ex)
+                {
+                    ex.Data["NON_REPORTED_RESOLVED_SPECIFICATIONS"] = result;
```

Comment:
Is there a reason not to add this as an explicit property on `AmbiguousArgumentsException`? (rather than in `Data`?). So we could do something like `ex.ResolvedSpecs = result` or `ex.AddAlreadyResolvedSpecs(result)`?

Alternatively the already resolved specs (`result`) could be passed to `_nonParamsArgumentSpecificationFactory.Create` and it could create an `AmbiguousArgumentsException` with the required info initialised already.

If it is still better to use `.Data[]` I think it would be worth extracting the string as a constant `AmbiguousArgumentsException.NonReportedResolvedSpecsDataKey` or similar (so the strings don't get out of sync between here and `MixedArgumentSpecificationsFactory`. (Ugh, so many factories 😭 !!!  😂)

Reply:
The reason why I don't want to add a property to the exception is that such information is very confusing to the end client. Currently this property doesn't contain all the resolved matchers (only a part of them) and I don't want to maintain this data to make it full and consistent more than we actually require. The usage of the `Data[]` dictionary greatly highlights a fact that data is very optional. Also I don't want users to even see that we provide this information.

The `.Data[]` property is a kind of hacky approach to carry additional information from the lower layers of system to the upper ones. In API those parts know nothing about each other, so test for the special dynamic data inside the exception extra payload looks like a kind of a bit clumsy very optional contract.

I'll introduce an internal constant as you suggested.